lib/graphql/metrics/analyzer.rb

```diff
@@ -0,0 +1,125 @@
+# frozen_string_literal: true
+
+module GraphQL
+  module Metrics
+    class Analyzer < GraphQL::Analysis::AST::Analyzer
+      attr_reader :query
+
+      def initialize(query_or_multiplex)
+        super
+
+        @query = query_or_multiplex
+        ns = query.context.namespace(CONTEXT_NAMESPACE)
+        ns[ANALYZER_INSTANCE_KEY] = self
+
+        @static_query_metrics = nil
+        @static_field_metrics = []
+      end
+
+      def analyze?
+        query.valid? && !query.context[GraphQL::Metrics::SKIP_GRAPHQL_METRICS_ANALYSIS]
+      end
+
+      def extract_query(runtime_query_metrics: {})
+        query_extracted(@static_query_metrics.merge(runtime_query_metrics)) if @static_query_metrics
+      end
+
+      def on_enter_operation_definition(_node, _parent, visitor)
+        @static_query_metrics = {
+          operation_type: visitor.query.selected_operation.operation_type,
+          operation_name: visitor.query.selected_operation.name,
+        }
+      end
+
+      def on_leave_field(node, _parent, visitor)
+        return if visitor.field_definition.introspection?
+        return if query.context[SKIP_FIELD_AND_ARGUMENT_METRICS]
+
+        # NOTE: @rmosolgo "I think it could be reduced to `arguments = visitor.arguments_for(ast_node)`"
```

Comment:
Do we need this note anymore?

Reply:
Yeah, I'm going to write up some GH issues for some tech debt stuff discovered along the way. I think this comment will be useful to leave here until then.